### PR TITLE
Avoid nullptr dereference during processing of NULL messages in Kafka for some formats

### LINKED ISSUE #19255: A bug in Kafka engine: Segmentation fault (JSONEachRowRowInputFormat)
Kafka ingestion crashed the server.
Data is serialized via jackson (possible missing field referenced  by queue/topic table)

```
2021.01.18 16:19:09.348955 [ 30912 ] {} <Fatal> BaseDaemon: ########################################
2021.01.18 16:19:09.349037 [ 30912 ] {} <Fatal> BaseDaemon: (version 20.8.3.18, no build id) (from thread 4994) (no query) Received signal Segmentation fault (11)
2021.01.18 16:19:09.349063 [ 30912 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.
2021.01.18 16:19:09.349096 [ 30912 ] {} <Fatal> BaseDaemon: Stack trace: 0x1131d9c0 0x118459ac 0x112719d7 0x11890535 0x11891774 0x10a7675d 0x10a91b8a 0x11472f8d 0x11473988 0x1149cde2 0x1149d20a 0x1149d30f 0xa43cc4d 0xa43b3ff 0x7fa73043bea5 0x7fa72fd5896d
2021.01.18 16:19:09.349160 [ 30912 ] {} <Fatal> BaseDaemon: 3. DB::JSONEachRowRowInputFormat::readPrefix() @ 0x1131d9c0 in /usr/bin/clickhouse
2021.01.18 16:19:09.349189 [ 30912 ] {} <Fatal> BaseDaemon: 4. DB::IRowInputFormat::generate() @ 0x118459ac in /usr/bin/clickhouse
2021.01.18 16:19:09.349212 [ 30912 ] {} <Fatal> BaseDaemon: 5. DB::ISource::work() @ 0x112719d7 in /usr/bin/clickhouse
2021.01.18 16:19:09.349229 [ 30912 ] {} <Fatal> BaseDaemon: 6. ? @ 0x11890535 in /usr/bin/clickhouse
2021.01.18 16:19:09.349246 [ 30912 ] {} <Fatal> BaseDaemon: 7. DB::KafkaBlockInputStream::readImpl() @ 0x11891774 in /usr/bin/clickhouse
2021.01.18 16:19:09.349264 [ 30912 ] {} <Fatal> BaseDaemon: 8. DB::IBlockInputStream::read() @ 0x10a7675d in /usr/bin/clickhouse
2021.01.18 16:19:09.349289 [ 30912 ] {} <Fatal> BaseDaemon: 9. DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::__1::atomic<bool>*) @ 0x10a91b8a in /usr/bin/clickhouse
2021.01.18 16:19:09.349307 [ 30912 ] {} <Fatal> BaseDaemon: 10. DB::StorageKafka::streamToViews() @ 0x11472f8d in /usr/bin/clickhouse
2021.01.18 16:19:09.349324 [ 30912 ] {} <Fatal> BaseDaemon: 11. DB::StorageKafka::threadFunc() @ 0x11473988 in /usr/bin/clickhouse
2021.01.18 16:19:09.349341 [ 30912 ] {} <Fatal> BaseDaemon: 12. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x1149cde2 in /usr/bin/clickhouse
2021.01.18 16:19:09.349358 [ 30912 ] {} <Fatal> BaseDaemon: 13. DB::BackgroundSchedulePool::threadFunction() @ 0x1149d20a in /usr/bin/clickhouse
2021.01.18 16:19:09.349373 [ 30912 ] {} <Fatal> BaseDaemon: 14. ? @ 0x1149d30f in /usr/bin/clickhouse
2021.01.18 16:19:09.349395 [ 30912 ] {} <Fatal> BaseDaemon: 15. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xa43cc4d in /usr/bin/clickhouse
2021.01.18 16:19:09.349411 [ 30912 ] {} <Fatal> BaseDaemon: 16. ? @ 0xa43b3ff in /usr/bin/clickhouse
2021.01.18 16:19:09.349435 [ 30912 ] {} <Fatal> BaseDaemon: 17. start_thread @ 0x7ea5 in /usr/lib64/libpthread-2.17.so
2021.01.18 16:19:09.349459 [ 30912 ] {} <Fatal> BaseDaemon: 18. __clone @ 0xfe96d in /usr/lib64/libc-2.17.so
```

Thank you for your effort and time.

Regards,
Ivan

### PR DESCRIPTION
**Pull request by filimonov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Prevent crashes for some formats when NULL (tombstone) message was coming from Kafka. Closes #19255.

Detailed description / Documentation draft:
As not all the parsers are ready to get nulls, we just skip them now.

For some formats (LineAsString, JSONAsSring, RawBLOB) theoretically we can do better (pass the NULL through as DB  NULL)  if the target is Nullable(String). But it's not clear how to do it correctly and if it's really needed.

### CODE REVIEW

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp [hunk 1]
```diff
@@ -38,6 +38,7 @@ ReadBufferFromKafkaConsumer::ReadBufferFromKafkaConsumer(
     , stopped(stopped_)
     , current(messages.begin())
     , topics(_topics)
+    , dummy_buf(new char[1])
```
Comment (alexey-milovidov) [inline src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp:41]: Why dynamic allocation?
Maybe simply `char dummy_buf`?

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp [hunk 2]
```diff
@@ -467,8 +468,26 @@ bool ReadBufferFromKafkaConsumer::nextImpl()
...
 
     // XXX: very fishy place with const casting.
-    auto * new_position = reinterpret_cast<char *>(const_cast<unsigned char *>(current->get_payload().get_data()));
-    BufferBase::set(new_position, current->get_payload().get_size(), 0);
+    auto message_data = current->get_payload().get_data();
+
+    char * new_position;
+    size_t message_size;
+
+    if (unlikely(message_data == NULL))
```
Comment (alexey-milovidov) [inline src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp:476]: nullptr

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp [hunk 3]
```diff
@@ -467,8 +468,26 @@ bool ReadBufferFromKafkaConsumer::nextImpl()
...
+    {
+        // in some cases message can be NULL (tombstone records for example)
+        // parsers are not ready to get NULLs on input, so the best we can do here
+        // it substitute NULL with an pointer to empty dummy buffer.
+        new_position = dummy_buf.get();
```
Comment (alexey-milovidov) [inline src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp:481]: It should be Ok to have empty buffer pointing to nullptr.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp (hunk 2 above)
Comment (filimonov) [reply src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp]: +

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp (hunk 1 above)
Comment (filimonov) [reply src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp]: +

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp (hunk 3 above)
Comment (filimonov) [reply src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp]: Sounds reasonable, but it is not always true (it happened not the first time, in older versions, other formats were also broken this way, see  #19255). Unless we have some automatic (unit?) tests checking that all the formats in the factory satisfy that invariant - we can't be sure. AFAIK that never happens outside of Kafka's use case, so it's easier to follow the behavior of other buffers.

Comment (filimonov) [reply src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp]: Also that makes the fix very local and simple to backport.

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp [hunk 4]
```diff
@@ -467,8 +467,26 @@ bool ReadBufferFromKafkaConsumer::nextImpl()
         return false;
 
     // XXX: very fishy place with const casting.
-    auto * new_position = reinterpret_cast<char *>(const_cast<unsigned char *>(current->get_payload().get_data()));
-    BufferBase::set(new_position, current->get_payload().get_size(), 0);
+    auto message_data = current->get_payload().get_data();
```
Comment (filimonov) [inline src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp:470]:
```suggestion
    auto * message_data = current->get_payload().get_data();
```

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.h [hunk 5]
```diff
@@ -99,6 +99,7 @@ class ReadBufferFromKafkaConsumer : public ReadBuffer
     // order is important, need to be destructed before consumer
     std::optional<cppkafka::TopicPartitionList> assignment;
     const Names topics;
+    char dummy_buf;
```
Comment (filimonov) [inline src/Storages/Kafka/ReadBufferFromKafkaConsumer.h:102]:
```suggestion
    char dummy_buf = 0;
```

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
This looks like a wrong change.

The reason is that ReadBuffer has invariant: on call of `next` is either must fill buffer with non-empty data or return false.

**Comment by filimonov** [comment]
> ReadBuffer has invariant

It does not exists/enforced in code, and Kafka buffer was  braking it even before that change (zero length messages)...  And there are even tests in Kafka suit checking empty messages. 

> must fill buffer with non-empty data or return false.

I have some doubts about that. Empty message = still a message. For ex. in protobuf it means 'record with all fields default'. I can try to change to return false, but i afraid that it may have nad side-effects (again the coverage by tests of Kafka vs all possible formats is not 100%).

For bug fix i think that approach is the safest.

File: tests/integration/test_storage_kafka/test.py [hunk 6]
```diff
@@ -283,6 +283,10 @@ def test_kafka_json_as_string(kafka_cluster):
     kafka_produce(kafka_cluster, 'kafka_json_as_string', ['{"t": 123, "e": {"x": "woof"} }', '', '{"t": 124, "e": {"x": "test"} }',
                                            '{"F1":"V1","F2":{"F21":"V21","F22":{},"F23":"V23","F24":"2019-12-24T16:28:04"},"F3":"V3"}'])
 
+    producer = KafkaProducer(bootstrap_servers="localhost:{}".format(cluster.kafka_port), value_serializer=producer_serializer, key_serializer=producer_serializer)
+    producer.send(topic='kafka_json_as_string', key='k9')
```
Comment (azat) [inline tests/integration/test_storage_kafka/test.py:287]: Worth a comment that this is a tombstone record?

Comment (filimonov) [reply tests/integration/test_storage_kafka/test.py:287]: +

**Comment by alexey-milovidov** [comment]
@filimonov
Returning false from `nextImpl` indicates end of stream. It is correct?
Do we read every message separately?

PS. The test passed but appeared to be flaky.

**Comment by alexey-milovidov** [comment]
> KeeperErrorCode = Session expired for /brokers/topics/consumer_hang2

**Comment by filimonov** [comment]
> @filimonov
> Returning false from `nextImpl` indicates end of stream. It is correct?

Yes. 

> Do we read every message separately? 

Yes. It is safe. 

> KeeperErrorCode = Session expired for /brokers/topics/consumer_hang2

We need either retries there (will check if i can add in the test itself, in other PR), either try to get rid of extra zookeepers there (reuse clickhouse - keeper, upgrade to newer Kafka which can work without zookeeper, or use redpanda)

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Reusing clickhouse-keeper for Kafka will be nice :)

**Pull request merged**